pkg/hostpath/controllerserver.go

```diff
@@ -136,3 +139,106 @@ func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req
 	}
 	return &csi.ValidateVolumeCapabilitiesResponse{Supported: true, Message: ""}, nil
 }
+
+func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		glog.V(3).Infof("invalid create snapshot req: %v", req)
+		return nil, err
+	}
+
+	// Check arguments
+	if len(req.GetSourceVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "SourceVolumeId missing in request")
+	}
+	if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Name missing in request")
+	}
+
+	volumeID := req.GetSourceVolumeId()
+	hostPathVolume, ok := hostPathVolumes[volumeID]
+	if !ok {
+		return nil, status.Error(codes.Internal, "volumeID is not exist")
+	}
+
+	snapshotID := uuid.NewUUID().String()
+	createAt := time.Now().UnixNano()
+	volPath := hostPathVolume.VolPath
+	file := snapshotRoot + snapshotID + ".tgz"
+	args := []string{"czf", file, "-C", volPath, "."}
+	executor := utilexec.New()
+	out, err := executor.Command("tar", args...).CombinedOutput()
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed create snapshot: %v: %s", err, out))
+	}
+
+	glog.V(4).Infof("create volume snapshot %s", file)
+	hostPathSna := hostPathSnapshot{}
+	hostPathSna.SnaName = req.GetName()
+	hostPathSna.SnaID = snapshotID
+	hostPathSna.VolID = volumeID
+	hostPathSna.SnaPath = file
+	hostPathSna.CreateAt = createAt
+	hostPathSna.Status = &csi.SnapshotStatus{
+		Type:    csi.SnapshotStatus_READY,
+		Details: fmt.Sprint("Successfully create snapshot"),
+	}
+
+	hostPathVolumeSnapshots[snapshotID] = hostPathSna
+
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			Id:             hostPathSna.SnaID,
+			SourceVolumeId: hostPathSna.VolID,
+			CreatedAt:      hostPathSna.CreateAt,
+			Status:         hostPathSna.Status,
+		},
+	}, nil
+}
+
+func (cs *controllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
+	// Check arguments
+	if len(req.GetSnapshotId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot ID missing in request")
+	}
+
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		glog.V(3).Infof("invalid delete snapshot req: %v", req)
+		return nil, err
+	}
+	snapshotID := req.SnapshotId
+	glog.V(4).Infof("deleting volume %s", snapshotID)
+	path := snapshotRoot + snapshotID + ".tgz"
+	os.RemoveAll(path)
+	delete(hostPathVolumeSnapshots, snapshotID)
+	return &csi.DeleteSnapshotResponse{}, nil
+}
+
+func (cs *controllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
+	// Check arguments
+	if len(req.GetSnapshotId()) == 0 {
```

Comment:
This is optional. It is only used to get the snapshot information about a single snapshot id. it should not be checked if it was passed.

Reply:
Also, we should return a list of all the snapshots in the system that:

1. Match the source volume id
2. Match the snapshot id
3. Match Anything (meaning all of the snapshots)